src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContextHelpers.cs

```diff
@@ -36,6 +43,14 @@ internal static void VerifyArguments(Diagnostic diagnostic, Func<Diagnostic, boo
             {
                 throw new ArgumentException(string.Format(CodeAnalysisResources.UnsupportedDiagnosticReported, diagnostic.Id), nameof(diagnostic));
             }
+
+            if (!diagnostic.Id.All(char.IsLetterOrDigit))
```

Comment:
I don't think we should make up new rules for valid identifiers. I suggest we use the same rules that the languages use (i.e. based on the Unicode specification plus underscore).